OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  /// argument has to be exactly the non-reference type of the metadata 
+  /// previously stored using insertMetaData(). The return value is a 
+  /// read-only reference to the metadata. Use updMetaData() to obtain a
+  /// writable reference. Time complexity is constant on average and linear in
+  /// number of elements in metadata on worst case.
+  ///
+  /// \throws OpenSim::MetaDataKeyDoesNotExist If the key specified does not
+  ///                                          exist in metadata.
+  /// \throws OpenSim::MetaDataTypeMismatch If the type specified as template
+  ///                                       argument does not match the type of
+  ///                                       metadata stored under the key
+  ///                                       specified.
+  template<typename ValueType>
+  const ValueType& getMetaData(const string& key) const {
+    static_assert(!std::is_reference<ValueType>::value, 
+                  "Template argument 'ValueType' should be exactly the" 
+                  " non-reference type of the MetaData value stored.");
+
+    try {
+      return metadata.at(key)->template getValue<ValueType>();
+    } catch(std::out_of_range&) {
+      std::string expl{"Key '" + std::string{key} + "' not found."};
+      throw MetaDataKeyDoesNotExist{expl};
+    } catch(std::bad_cast&) {
+      throw MetaDataTypeMismatch{"Template argument specified for getMetaData"
+                                 " is incorrect."};
+    }
+  }
+
+  /// Get previously inserted metadata using its key and type. The template
+  /// argument has to be exactly the non-reference type of the metadata 
+  /// previously stored using insertMetaData(). The returned value is editable.
+  /// Time complexity is constant on average and linear in number of elements
+  /// in metadata on worst case.
+  ///
+  /// \throws OpenSim::MetaDataKeyDoesNotExist If the key specified does not
+  ///                                          exist in metadata.
+  /// \throws OpenSim::MetaDataTypeMismatch If the type specified as template
+  ///                                       argument does not match the type of
+  ///                                       metadata stored under the key
+  ///                                       specified.
+  template<typename ValueType>
+  ValueType& updMetaData(const string& key) {
+    static_assert(!std::is_reference<ValueType>::value, "Template argument " 
+                  "'ValueType' should be exactly the non-reference type of"
+                  "the MetaData value stored.");
+
+    try {
+      return metadata.at(key)->template updValue<ValueType>();
+    } catch(std::out_of_range&) {
+      std::string expl{"Key '" + std::string{key} + "' not found."};
+      throw MetaDataKeyDoesNotExist{expl};
+    } catch(std::bad_cast&) {
+      throw MetaDataTypeMismatch{"Template argument specified for updMetaData"
+                                 " is incorrect"};
+    }
```

Comment:
> Could SimTK::NiceTypeName be used here? 

That's a good idea. `SimTK::NiceTypeName<T>::namestr()` should make a human-readable demangled type name on all platforms.